File: src/hal/gpio.c
#include "hal/gpio.h"

#include "hal/pinutils.h"
#include "hal/rcc.h"

#include <stdbool.h>
#include <stdint.h>

void gpio_set_mode(uint16_t const pin, uint8_t const mode)
{
    gpio_t *gpio = GPIO(PINBANK(pin));
    uint32_t n = PINNO(pin);
    RCC->AHB1ENR |= BIT(PINBANK(pin));
    gpio->MODER &= ~(0x3U << (n * 2));
    gpio->MODER |= (mode & 0x3U) << (n * 2);
}

bool gpio_set_af(uint16_t const pin, uint8_t const af_num)
{
    gpio_t *gpio = GPIO(PINBANK(pin));
    int n = PINNO(pin);
    gpio->AFR[n >> 3] &= ~(15UL << ((n & 7) * 4));
    gpio->AFR[n >> 3] |= ((uint32_t)af_num) << ((n & 7) * 4);
    return true;
}

void gpio_write(uint16_t const pin, bool const val)
{
    gpio_t *gpio = GPIO(PINBANK(pin));
    if (val) {
        /* Setting using BSRR enables atomic setting of the ODR register, which is
         * the output data register for the gpio */
        /* gpio->ODR |= (1U << PINNO(pin)); */
        gpio->BSRR |= (1U << PINNO(pin));
    } else {
        /* turning off a gpio is done by setting a "reset" bit which is 16 bits
         * further left than the pin number */
        /* gpio->ODR &= ~(1U << PINNO(pin)); */
        gpio->BSRR |= (1U << PINNO(pin) << 16U);
    }
}

File: src/zig/hal/gpio.zig
const c = @cImport({
    @cInclude("hal/gpio.h");
});

const Mode = enum(u8) {
    input = 0,
    output = 1,
    af = 2,
    analog = 3,
};

pub fn setMode(pin: u16, mode: Mode) void {
    c.gpio_set_mode(pin, @intFromEnum(mode));
}

pub fn setAltFunc(pin: u16, af: u3) bool {
    return c.gpio_set_af(pin, af);
}

pub fn write(pin: u16, value: bool) void {
    c.gpio_write(pin, value);
}
